contracts/strategies/BaseStrategy.sol

```diff
@@ -0,0 +1,217 @@
+        emit LogToggleStrategyExecutor(executor, strategyExecutors[executor]);
+    }
+
+    /// @inheritdoc IStrategy
+    function skim(uint256 amount) external override {
+        _skim(amount);
+    }
+
+    /// @notice Invests the underlying asset.
+    /// @param amount The amount of tokens to invest.
+    /// @dev Assume the contract's balance is greater than the amount
+    function _skim(uint256 amount) internal virtual;
+
+    /// @notice Harvest profits while preventing a sandwich attack exploit.
+    /// @param maxBalance The maximum balance of the underlying token that is allowed to be in BentoBox.
+    /// @param rebalance Whether BentoBox should rebalance the strategy assets to acheive it's target allocation.
+    /// @param maxChangeAmount When rebalancing - the maximum amount that will be deposited to or withdrawn from a strategy.
+    /// @param harvestRewards Whether we should also harvest any rewards the strategy has.
+    /// @dev maxBalance can be set to 0 to keep the previous value.
+    /// @dev maxChangeAmount can be set to 0 to allow for full rebalancing.
+    /// @dev Only executors can call safeHarvest on this strategy.
+    function safeHarvest(
+        uint256 maxBalance,
+        bool rebalance,
+        uint256 maxChangeAmount,
+        bool harvestRewards
+    ) external onlyExecutor {
+        if (harvestRewards) _harvestRewards();
+        if (maxBalance > 0) maxBentoBoxBalance = maxBalance;
+        bentoBox.harvest(address(underlying), rebalance, maxChangeAmount);
+    }
+
+    /// @inheritdoc IStrategy
+    /// @dev Only BentoBox can call harvest on this strategy.
+    function harvest(uint256 balance, address sender) external override onlyBentobox returns (int256) {
+        /** @dev Ensures that (1) the caller was this contract (called through the safeHarvest function)
+		        and (2) that we are not being frontrun by a large BentoBox deposit when harvesting profits.
+		    @dev Don't revert if conditions aren't met in order to allow
+		        BentoBox to continiue execution as it might need to do a rebalance. */
+        if (sender == address(this) && IBentoBoxMinimal(bentoBox).totals(address(underlying)).elastic <= maxBentoBoxBalance) {
+            /** @dev We might also have some underlying tokens in the contract from before so the amount
+                returned by the internal _harvest function isn't necessary the final profit/loss amount */
+            int256 amount = _harvest(balance);
+
+            uint256 contractBalance = IERC20(underlying).safeBalanceOf(address(this));
+
+            if (amount >= 0) {
+                // we made some profit
+                if (contractBalance > 0) IERC20(underlying).safeTransfer(address(bentoBox), contractBalance);
+                return int256(contractBalance);
+            } else if (contractBalance > 0) {
+                // we might have made a loss
+                int256 diff = amount + int256(contractBalance);
```

Comment:
Why would we sum the amount and contractBalance if a loss is already reported?
Shouldn't anything sitting on the strategy contract be accounted for?